chain/pra/receiver_relay.go

```diff
@@ -0,0 +1,256 @@
+					}
+
+					r.log.Debugf("findParasInclusionCandidateIncludedHead: found at relayblock %d", header.Number)
+					return header, &blockHash
+				}
+			}
+		}
+
+		blockNumber++
+	}
+
+	return nil, nil
+}
+
+func (r *relayReceiver) buildBlockUpdates(nexMtaHeight uint64, gj *substrate.GrandpaJustification, fetchtedBlockHeaders []substrate.SubstrateHeader) ([][]byte, error) {
+	bus := make([][]byte, 0)
+	blockHeaders := make([]substrate.SubstrateHeader, 0)
+
+	// Fetch headers with fetched blockheaders.
+	from := nexMtaHeight
+	to := uint64(gj.Commit.TargetNumber)
+
+	if len(fetchtedBlockHeaders) > 0 {
+		from = uint64(fetchtedBlockHeaders[len(fetchtedBlockHeaders)-1].Number)
+
+		// not need to fetch again
+		if from == to {
+			to -= 1
+		}
+
+		nextMtaHash, err := r.c.GetBlockHash(nexMtaHeight)
+		if err != nil {
+			return nil, err
+		}
+
+		nextMtaBlockHeader, err := r.c.GetHeader(nextMtaHash)
+		if err != nil {
+			return nil, err
+		}
+
+		blockHeaders = append(blockHeaders, *nextMtaBlockHeader)
+	}
+
+	misisingBlockNumbers := make([]substrate.SubstrateBlockNumber, 0)
+
+	for i := from; i <= to; i++ {
+		misisingBlockNumbers = append(misisingBlockNumbers, substrate.SubstrateBlockNumber(i))
+	}
+
+	missingBlockHeaders, err := r.c.GetBlockHeaderByBlockNumbers(misisingBlockNumbers)
+	if err != nil {
+		return nil, err
+	}
+
+	blockHeaders = append(blockHeaders, fetchtedBlockHeaders...)
+	blockHeaders = append(blockHeaders, missingBlockHeaders...)
+
+	for i, blockHeader := range blockHeaders {
+		var bu []byte
+		var votes []byte = nil
+		if i == len(blockHeaders)-1 {
+			votes, err = r.newVotes(gj, substrate.SetId(r.expectSetId))
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		bu, err = r.newBlockUpdate(blockHeader, votes)
+		if err != nil {
+			return nil, err
+		}
+
+		r.log.Tracef("buildBlockUpdates: %d", blockHeader.Number)
+
+		// Sync MTA
+		r.updateMta(uint64(blockHeader.Number-1), blockHeader.ParentHash)
+		r.expectMtaHeight = uint64(blockHeader.Number - 1)
+
+		bus = append(bus, bu)
+	}
+
+	// Sync MTA
+	relayBlockhash, err := r.c.GetBlockHash(uint64(blockHeaders[len(blockHeaders)-1].Number))
+	if err != nil {
+		return nil, err
+	}
+
+	r.updateMta(uint64(blockHeaders[len(blockHeaders)-1].Number), relayBlockhash)
+	r.expectMtaHeight = uint64(blockHeaders[len(blockHeaders)-1].Number)
+
+	r.log.Debugf("buildBlockUpdates: %d ~ %d", blockHeaders[0].Number, blockHeaders[len(blockHeaders)-1].Number)
+	return bus, nil
+}
+
+func (r *relayReceiver) buildFinalityProof(includeHeader *substrate.SubstrateHeader, includeHash *substrate.SubstrateHash) ([][]byte, error) {
+	finalityProofs := make([][]byte, 0)
+
+	if r.expectMtaHeight < uint64(r.bmvStatus.RelayMtaHeight) {
+		r.expectMtaHeight = uint64(r.bmvStatus.RelayMtaHeight)
+		r.expectSetId = uint64(r.bmvStatus.RelaySetId)
+	}
+
+	// For blockproof to work
+	for r.expectMtaHeight < uint64(includeHeader.Number) {
+		bus := make([][]byte, 0)
```

Comment:
Do we need to declare this variable? 
Shouldn't we name it as `bus` this may cause confusion. ?
`blockupdates` is not bad to me.

blockupdates, err := r.buildBlockUpdates(uint64(r.expectMtaHeight+1), gj, bhs)

also remove
```
bus = append(bus, newBus...)
```